Initial commit for ML ops guide (#12930)
dagster-io/dagster: docs/content/guides/dagster/ml-ops.mdx

---

> title: Building machine learning pipelines In Dagster | Dagster Docs

Reviewer: "In" should not be capitalized

description: This guide illustrates how to use Dagster to operationalize your machine learning pipeline
---

# Building machine learning pipelines with Dagster

In this guide, we’ll walk you through how to take your machine learning models and deploy and maintain them in production using Dagster, reliably and efficiently.

> We will work through building a machine lerning pipeline, including using assets for different elements, how to automate model training, and monitoring your model's drift.

Reviewer: ```suggestion
We will work through building a machine learning pipeline, including using assets for different elements, how to automate model training, and monitoring your model's drift.
```

<Note>
  This guide assumes you have familiarity with machine learning concepts and
  several Dagster concepts, including{" "}
  <a href="/concepts/assets/software-defined-assets">software-defined assets</a>{" "}
  and <a href="/concepts/ops-jobs-graphs/jobs">jobs</a> . We also recommend
  reviewing the{" "}
  <a href="/guides/dagster/automating-pipelines">
    Automating Data Pipelines guide
  </a>{" "}
  as many of the same principles can be applied to machine learning.
</Note>

---

## Machine learning development

If you are already using Dagster for your ETL pipelines, it is a natural progression to build out and test your models in Dagster.

For this guide, we will be using the Hacker News data demoed in the [tutorial](/tutorial).

The machine learning model we will walk through takes the Hacker News stories and uses the titles to predict the number of comments that a story will generate. This will be a supervised model since we have the number of comments for all the previous stories.

The assets graph will look like this at the end of this guide:

<Image
alt="alt"

> src="/images/guides/ml-ops/ml_asset_dag.png"

Author: that should be better!

width={1804}
height={1064}
/>

### Ingesting Data

First, we will create an asset that retrieves the most recent Hacker News records.

```python file=/guides/dagster/ml_ops/ml_ops.py startafter=data_ingestion_start endbefore=data_ingestion_end
import requests
from dagster import asset, FreshnessPolicy
import pandas as pd

> @asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=10))

Reviewer: I think we should probably actually leave out the `FreshnessPolicy`, at least for now. Because we're not telling people how to schedule their model, so its presence is likely to be confusing.

I also think they might not necessarily want to update their training data every 10 minutes.

Author: ok removed from this and inference

    latest_item = requests.get(
        f"https://hacker-news.firebaseio.com/v0/maxitem.json"
    ).json()
    # Get items based on story ids from the HackerNews items endpoint
    results = []
    scope = range(latest_item - 100, latest_item)
    for item_id in scope:
        item = requests.get(
            f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
        ).json()
        results.append(item)

    df = pd.DataFrame(results)
    if len(df) > 0:
        df = df[df.type == "story"]
        df = df[~df.title.isna()]
    inference_x = df.title
    # Transform the new story titles using the existing vectorizer
    inference_x = Tfidf_Vectorizer.transform(inference_x)
    return xgboost.predict(inference_x)
```

Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history, and trigger retraining.

---

## Where to go from here

- Dagster is flexible so there could be many 'right' ways to implement your machine learning pipeline and different avenues to explore
- Dagster intergrates with [MLflow](/\_apidocs/libraries/dagster-mlflow) that can be used to keep track of your models

> - Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.

Reviewer: A backslash is showing up when this is rendered

Reviewer 2: This is still an issue for me as well

---
title: Building your machine learning pipelines In Dagster | Dagster Docs
description: This guide illustrates how to use Dagster to operationaize your machine learning pipeline
---

> # Building your machine learning pipelines with Dagster

Reviewer: I would take out "your"

Reviewer 2: +1

In this guide, we’ll walk you through how to take your machine learning models and deploy and mantain them in production using Dagster, reliably and efficently.

<Note>
  This guide assumes you have familiarity with machine learning concepts,{" "}

>   <a href="/tutorial">the Dagster tutorial</a> and several Dagster concepts,

Reviewer: I don't think we necessarily assume they have familiarity with the tutorial, just with the concepts. We can suggest that they take the tutorial if they're not familiar with the concepts.

Reviewer 2: As this guide builds on the tutorial, that's why there's familiarity suggested here. I agree we can rephrase though. Perhaps something along these lines?:

> This guide assumes you have familiarity with machine learning and several Dagster concepts. It also builds on the [asset tutorial](), which we recommend completing if you're unfamiliar with Dagster or want a comprehensive look at the project code.

Reviewer: +1

  including{" "}
  <a href="/concepts/assets/software-defined-assets">software-defined assets</a>{" "}
  and <a href="/concepts/ops-jobs-graphs/jobs">jobs</a> . We also recommend
  reviewing the{" "}
  <a href="/guides/dagster/automating-pipelines">
    automating data pipelines guide
  </a>{" "}
  as many of the same principles can be applied to MLOps.
</Note>

> We will work through several aspects of MLOps, including using assets for different elements of your machine learning pipeline, how to automate model training and monitoring your model's drift.

Reviewer: Oxford comma plz

---

## Why build your ML pipelines with Dagster?

- Dagster is designed to use during development, the lightweight execution model means you can access the benefits of an orchestrator, like re-executing from the middle of a pipeline and parallelizing steps while you're experiementing.

> - It models data assets, not just tasks so Dagster understands the upstream and downstream data depenedencies and is a one-stop shop for both the data transformations and the models that depend on the data transformations.

Reviewer: ```suggestion
- Dagster models data assets, not just tasks, so it understands the upstream and downstream data dependencies and is a one-stop shop for both the data transformations and the models that depend on the data transformations.
```

Reviewer: This is a bit of a run-on sentence.

---

## Machine learning development

> Before we go into how to operatialize a machine learning model, let's go over some of the tools you can use in Dagster to develop a machine learning model.

Reviewer: What tools do you mean?

If you are already using Dagster for your ETL pipelines, it is a natural progression to build out and test your models in Dagster.

For this guide, we will be using the hackernews data which is demoed in the [tutorial](/tutorial).

The machine learning model we will walk through is taking the hackernews stories, using the title to predict the number of comments that a story will generate. This will be a supervised model since we have the number of comments for all the previous stories.

The assets graph will look like this at the end of this guide:

<Image
alt="alt"
src="/images/guides/ml-ops/ml_asset_dag.png"
width={1804}
height={1064}
/>

### Data Ingestion

> First, we will be creating an asset which retrieves all the hackernews records from the previous run to the latest hackernews record.

Reviewer: This isn't exactly accurate, correct? In the sense that we're not retrieving all the records from the previous run?

Author: updated

```python file=/guides/dagster/ml_ops/ml_ops.py startafter=data_ingestion_start endbefore=data_ingestion_end
import requests
from dagster import asset, FreshnessPolicy
import pandas as pd

@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=10))
def hackernews_stories():
    """Get the max ID number from hacker news"""
    latest_item = requests.get(
        f"https://hacker-news.firebaseio.com/v0/maxitem.json"
    ).json()
    """Get items based on story ids from the HackerNews items endpoint"""
    results = []
    scope = range(latest_item-1000, latest_item)
    for item_id in scope:
        item = requests.get(
            f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
        ).json()
        results.append(item)

    df = pd.DataFrame(results)
    if len(df) > 0:
        df = df[df.type == "story"]
        df = df[~df.title.isna()]

    return df
```

### Data Transformation

Now that we have a dataframe with all valid stories, we want to transform that data into something our machine learning model will be able to use.

> The first step is taking the dataframe and splitting into a [training and test set](https://en.wikipedia.org/wiki/Training,\_validation,\_and_test_data_sets). In some of your models, you also might choose to have an additional split for a validation set. The reason we split the data is so that we can have a test and/or a validation dataset that is independent of the training set and we can use that to see how well our model did.

Reviewer 2: ```suggestion
The first step is taking the dataframe and splitting into a [training and test set](https://en.wikipedia.org/wiki/Training,\_validation,\_and_test_data_sets). In some of your models, you also might choose to have an additional split for a validation set. The reason we split the data is so that we can have a test and/or a validation dataset that is independent of the training set. We can then use that dataset to see how well our model did.
```

🔧 Suggesting you break up the last sentence in this paragraph. It currently reads a bit like a run on and can be difficult to follow.

    """Initiate and fit the tokenizer on the training data and transform the training dataset"""
    vectorizer = TfidfVectorizer()
    transformed_X_train = vectorizer.fit_transform(X_train)
    transformed_X_train = transformed_X_train.toarray()
    y_train = y_train.fillna(0)
    transformed_y_train = np.array(y_train)
    return vectorizer, (transformed_X_train, transformed_y_train)

@asset
def transformed_test_data(test_data, Tfidf_Vectorizer):
    X_test, y_test = test_data
    """Use the fitted tokenizer to transform the test dataset"""
    transformed_X_test = Tfidf_Vectorizer.transform(X_test)
    transformed_y_test = np.array(y_test)
    y_test = y_test.fillna(0)
    transformed_y_test = np.array(y_test)
    return transformed_X_test, transformed_y_test
```

We have also transformed the dataframes into numpy arrays and removed na values to prepare the data for training.

### Model training

At this point, we have X_train, y_train, X_test, y_test ready to go for our model. We can use any number of models to train our model from libraries like [sklearn](https://scikit-learn.org/), [tensorflow](https://www.tensorflow.org/) and [pytorch](https://pytorch.org/).

> In our example, we will be training a [XGBoost model](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor) to predict a numerical value.

Reviewer 2: ```suggestion
In our example, we will train an [XGBoost model](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor) to predict a numerical value.
```

🔧 Suggesting we use `we will train` instead of `we will be training` - it's less wordy and more action-forward.

```python file=/guides/dagster/ml_ops/ml_ops.py startafter=models_start endbefore=models_end
import xgboost as xg
from sklearn.metrics import mean_absolute_error

@asset
def xgboost(transformed_training_data):
    transformed_X_train, transformed_y_train = transformed_training_data
    """Train XGBoost model, which is a highly efficent and flexible model"""
    xgb_r = xg.XGBRegressor(objective ='reg:squarederror', eval_metric=mean_absolute_error,
                  n_estimators = 20)
    xgb_r.fit(transformed_X_train, transformed_y_train)
    return xgb_r

@asset
def score_xgboost( transformed_test_data, xgboost):
    transformed_X_test, transformed_y_test = transformed_test_data
    """Use the test set data to get a score of the XGBoost model"""
    score = xgboost.score(transformed_X_test, transformed_y_test)
    return score
```

> We include freshness policies in the models here which will ensure the upstream data transformation assets are updated to comply with these policies.

Reviewer 2: ❓ Include a link to freshness policy documentation?

### Evaluate our results

> In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results. Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.

Reviewer 2: > In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results.

What error? I wasn't instructed to run or look at anything, so how can I know there's an error?

> Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.

I'm finding this hard to follow - I think it's because of how it's structured. Is this accurate?:

> Next, to predict the results, we'll create another asset that runs inference on the model more frequently than the model is re-trained.

Author: updated- do we know how to put R^2 in nicely?

```python file=/guides/dagster/ml_ops/ml_ops.py startafter=inference_start endbefore=inference_end
@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=60))
def model_inference(xgboost, Tfidf_Vectorizer):
    """Get the max ID number from hacker news"""

    latest_item = requests.get(
        f"https://hacker-news.firebaseio.com/v0/maxitem.json"
    ).json()
    """Get items based on story ids from the HackerNews items endpoint"""
    results = []
    scope = range(latest_item-100, latest_item)
    for item_id in scope:
        item = requests.get(
            f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
        ).json()
        results.append(item)

    df = pd.DataFrame(results)
    if len(df) > 0:
        df = df[df.type == "story"]
        df = df[~df.title.isna()]
    inference_x = df.title
    inference_x = Tfidf_Vectorizer.transform(inference_x)
    return xgboost.predict(inference_x)
```

> Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history and trigger retraining.

Reviewer 2: ```suggestion
Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history, and trigger retraining.
```

🔧 Oxford comma pls

> Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.

Author: was between scoring and predicting here..

> Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.

Reviewer 2: 🤔 What's the user supposed to do with this information? Use W&B in their own models? Currently, it feels a bit tacked on - it could use some additional info.

Author: added more of a section with some other recommendations

---

> ## Machine learning operations

Reviewer: My instinct would be to leave this out for the initial PR.  I think there's value in this kind of content, and I think this is a solid start, but I also think it will take some back and forth to get it right, and I'd rather not have this block the rest of the guide from making it in.

Author: sounds good saving it in notion for later
